### PR TITLE
[platform] Support PDU reboot when DUT crash after OOM reboot

### PR DESCRIPTION
**Pull request by lizhijianrd**
### Description of PR

Summary:
Some DUTs crashed on test case `platform_tests.test_memory_exhaustion`. We can try to recover the DUT by PDU reboot. In this PR, we add PDU reboot to the `teardwon` phase.

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 201911
- [x] 202012

#### What is the motivation for this PR?
Some DUTs crashed on test case `platform_tests.test_memory_exhaustion`. We can try to recover the DUT by PDU reboot. In this PR, we add PDU reboot to the `teardwon` phase.

#### How did you do it?
Add PDU reboot to the `teardwon` phase.

#### How did you verify/test it?
Tested on physical testbeds.

### CODE REVIEW

File: tests/platform_tests/test_memory_exhaustion.py [hunk 1]
```diff
@@ -13,48 +14,79 @@
...
+                                 port=SONIC_SSH_PORT,
+                                 state=expected_state,
+                                 search_regex=SONIC_SSH_REGEX,
+                                 delay=10,
+                                 timeout=timeout,
+                                 module_ignore_errors=True)
+        return not res.is_failed and 'Timeout' not in res.get('msg', '')
+
+    def pdu_reboot(self, pdu_controller):
+        pdu_controller.turn_off_outlet()
```
Comment (Blueve) [inline tests/platform_tests/test_memory_exhaustion.py:90]: what happens if the pdu controller doesn't exists or the control command failed?

Comment (lizhijianrd) [reply tests/platform_tests/test_memory_exhaustion.py]: I checked if the PDU controller exists at line 34. I added some code to check the results of these control commands in a new commit.

File: tests/platform_tests/test_memory_exhaustion.py [hunk 2]
```diff
@@ -13,48 +14,79 @@
...
+
+    @pytest.fixture(autouse=True)
+    def teardown(self, duthost, localhost, pdu_controller):
+        yield
+        # If the SSH connection is not established, or any critical process is exited,
+        # try to recover the DUT by PDU reboot.
+        dut_ip = duthost.mgmt_ip
+        hostname = duthost.hostname
+        if not self.check_ssh_state(localhost, dut_ip, SSH_STATE_STARTED):
+            if pdu_controller is None:
```
Comment (lizhijianrd) [inline tests/platform_tests/test_memory_exhaustion.py:34]: @Blueve I check if the pdu controller exists at here.

File: tests/platform_tests/test_memory_exhaustion.py [hunk 3]
```diff
@@ -13,48 +14,83 @@
...
+                                 state=expected_state,
+                                 search_regex=SONIC_SSH_REGEX,
+                                 delay=10,
+                                 timeout=timeout,
+                                 module_ignore_errors=True)
+        return not res.is_failed and 'Timeout' not in res.get('msg', '')
+
+    def pdu_reboot(self, pdu_controller):
+        hostname = pdu_controller.dut_hostname
+        if not pdu_controller.turn_off_outlet():
```
Comment (lizhijianrd) [inline tests/platform_tests/test_memory_exhaustion.py:91]: @Blueve I added some code to check the result of these control commands.

**Review by Blueve — APPROVED** [review]

**Pull request merged**